mesonbuild/modules/pkgconfig.py

```diff
@@ -329,6 +330,39 @@ def _generate_pkgconfig_file(self, state, deps, subdirs, name, description,
                                  url, version, pcfile, conflicts, variables,
                                  unescaped_variables, uninstalled=False, dataonly=False):
         coredata = state.environment.get_coredata()
+        referenced_vars = set()
+        optnames = [x.name for x in BUILTIN_DIR_OPTIONS.keys()]
+
+        if not dataonly:
+            # includedir is always implied, although libdir may not be
+            # needed for header-only libraries
+            referenced_vars |= {'prefix', 'includedir'}
+            if deps.pub_libs or deps.priv_libs:
+                referenced_vars |= {'libdir'}
+        # also automatically infer variables referenced in other variables
+        implicit_vars_warning = False
+        redundant_vars_warning = False
+        varnames = set()
+        varstrings = set()
+        for k, v in variables + unescaped_variables:
+            varnames |= {k}
+            varstrings |= {v}
+        for optname in optnames:
+            optvar = f'${{{optname}}}'
+            if any(x.startswith(optvar) for x in varstrings):
+                if optname in varnames:
+                    redundant_vars_warning = True
+                else:
+                    implicit_vars_warning = True
+                    referenced_vars |= {'prefix', optname}
+        if redundant_vars_warning:
+            FeatureDeprecated.single_use('pkgconfig.generate variable for builtin directories', '0.62.0',
+                                         state.subproject, 'They will be automatically included when referenced',
+                                         state.current_node)
+        if implicit_vars_warning:
+            FeatureNew.single_use('pkgconfig.generate implicit variable for builtin directories', '0.62.0',
+                                  state.subproject, location=state.current_node)
+
```

Comment:
This isn't entirely accurate if libdir is the only implicit variable. For example, the build of the gst-plugins-base package that I mentioned in #9992 now complains loudly multiple times, yet prior to the breaking commit this was just fine.

That said, this works fine and is a nice fix and given that this is only a warning I don't think it's worth being all that nitpicky about it, especially since the extra check needed would complicate things.

Reply:
Oh right, that is an annoying confluence of edge cases indeed.

Reply:
I think this is a valid point and not really an edge case. prefix, libdir and includedir were always guaranteed to be part of the generated file, so using them implicitly is not a new behaviour.

Reply:
Alright, I added a suppression for this, and a test case that checks both that the common case is warned about, and that your case is silent.

Reply:
Also yes it is absolutely an edge case. ***Edge cases should be handled, not left for people to cut themselves on***, which is why I was always going to fix this and just needed to get home and work on it... my comment simply meant that "hmm, I can see why I overlooked that".

Reply:
oh ok, I understood your comment as it's rare enough that we could leave it like that.